src/clients/dispatcher/heartbeat/heartbeat-worker.ts

```diff
@@ -0,0 +1,89 @@
+import { parentPort, workerData } from 'worker_threads';
+import { Logger } from '@util/logger';
+import { ClientConfig, addTokenMiddleware, channelFactory } from '@hatchet/clients/hatchet-client';
+import { DispatcherClient as PbDispatcherClient } from '@hatchet/protoc/dispatcher';
+import { ConfigLoader } from '@hatchet/util/config-loader';
+import { Status, createClientFactory } from 'nice-grpc';
+import { DispatcherClient } from '../dispatcher-client';
+
+const HEARTBEAT_INTERVAL = 4000;
+
+class HeartbeatWorker {
+  heartbeatInterval: any;
+  logger: Logger;
+  client: PbDispatcherClient;
+  workerId: string;
+  timeLastHeartbeat = new Date().getTime();
+
+  constructor(config: ClientConfig, workerId: string) {
+    this.workerId = workerId;
+
+    this.logger = new Logger(`Heartbeat`, config.log_level);
+
+    const credentials = ConfigLoader.createCredentials(config.tls_config);
+    const clientFactory = createClientFactory().use(addTokenMiddleware(config.token));
+
+    const dispatcher = new DispatcherClient(
+      config,
+      channelFactory(config, credentials),
+      clientFactory
+    );
+
+    this.client = dispatcher.client;
+  }
+
+  async start() {
+    if (this.heartbeatInterval) {
+      return;
+    }
+
+    const beat = async () => {
+      try {
+        this.logger.debug('Heartbeat sending...');
+        await this.client.heartbeat({
+          workerId: this.workerId,
+          heartbeatAt: new Date(),
+        });
+        const now = new Date().getTime();
+
+        const actualInterval = now - this.timeLastHeartbeat;
+
+        if (actualInterval > HEARTBEAT_INTERVAL * 1.2) {
+          this.logger.warn(
+            `Heartbeat interval delay (${actualInterval}ms >> ${HEARTBEAT_INTERVAL}ms)`
+          );
+        }
+
+        this.logger.debug(`Heartbeat sent ${actualInterval}ms ago`);
+        this.timeLastHeartbeat = now;
+      } catch (e: any) {
+        if (e.code === Status.UNIMPLEMENTED) {
+          // break out of interval
+          this.logger.error('Heartbeat not implemented, closing heartbeat');
+          this.stop();
```

Comment:
Do we leak a thread here if we call `this.stop()` without the parent calling `terminate` on the worker?

Reply:
yes, but there's no way to terminate the worker thread from with in the worker thread
we should never be calling HeartbeatWorker.stop() directly, instead calling stop from the controller.